### PR TITLE
Format type aliases

### LINKED ISSUE #486: Format type aliases

### PR DESCRIPTION
**Pull request by Marwes**
Fixes #486

### CODE REVIEW

**Comment by Marwes** [comment]
A bit bigger change than the small fixes before so some also looking to hear if everything is named and located correctly. It might need some more tests as well but I wanted some feedback before venturing further.

File: src/items.rs [hunk 1]
```diff
@@ -638,6 +638,46 @@ fn format_tuple_struct(context: &RewriteContext,
...
+                          span: Span)
+                          -> Option<String> {
+    let mut result = String::new();
+
+    result.push_str(&format_visibility(vis));
+    result.push_str("type ");
+    result.push_str(&ident.to_string());
+
+    let generics_indent = indent + result.len();
+    let generics_span = mk_sp(span.lo, ty.span.lo);
```
Comment (nrc) [inline src/items.rs]: You should use utils::span_after(... "type") to get a more precise start for the generics span here

File: src/items.rs [hunk 2]
```diff
@@ -638,6 +638,46 @@ fn format_tuple_struct(context: &RewriteContext,
...
+    let generics_str = try_opt!(format_generics(context,
```
Comment (nrc) [inline src/items.rs]: You might be better off using `rewrite_generics` here - there is no where clause so (I think) using `format_generics` is overkill.

File: tests/target/type_alias.rs [hunk 3]
```diff
@@ -0,0 +1,7 @@
+
+type PrivateTest<'a, I> = (Box<Parser<Input = I, Output = char> + 'a>,
+                           Box<Parser<Input = I, Output = char> + 'a>);
+
+pub type PublicTest<'a, I, O> = Result<Vec<MyLongType>,
+                                       Box<Parser<Input = I, Output = char> + 'a>,
+                                       Box<Parser<Input = I, Output = char> + 'a>>;
```
Comment (nrc) [inline tests/target/type_alias.rs]: I would like to see some more tests - in particular tests with comments inside the generics and where long type names mean the generics splits over multiple lines. Also a test where the result fits in exactly 100 chars to ensure it ends up on one line.

**Comment by nrc** [comment]
Thanks! This looks great, a few more tests would be good and there are some very minor improvements you could make (see comments inline). r+ with those things addressed.

File: src/items.rs (hunk 2 above)
Comment (Marwes) [inline src/items.rs]: So the `rewrite_*` functions are more "lightweight" versions for the `format_` functions?

Comment (nrc) [inline src/items.rs]: Not always, format_generics and rewrite_generics is really bad naming tbh, we should change it. (There is history with the rewrite_* and format_* names, but I don't think there is any real distinction any more).

**Comment by Marwes** [comment]
Haven't been able to work much on this but I added tests for long parameter lists and a test for comments. It appears that only comments before type parameters are saved however.

Long parameter lists are formatted but how should it be formatted when the type cannot fit on the end of the line?

```
//This does not fit
pub type ExtremelyLongName<
                                                 And,
                                                 LongParameters> = LongLongType;
//Format it like this
pub type ExtremelyLongName<
                                                 And,
                                                 LongParameters>
    = LongLongType;
//Prefer the entire parameter list be moved to a new line first?
pub type ExtremelyLongNameAndParameters
     <And, LongParameters> = LongLongType;
```

I will rebase this on top of master after this is resolved so it can be merged.

**Comment by nrc** [comment]
Line breaking of generics should take care of itself, if you still need to break the line, I would do so after the `=` and nowhere else. It would be better to break after the `=` and not break in the generics, if it is possible to do that.

**Comment by Marwes** [comment]
I think I fixed everything you mentioned now.

I noticed that comments are not preserved between the '=' and the type but they seem to be lost before function return types as well so that might warrant a separate PR to deal with that before any start of a type.

File: src/items.rs [hunk 4]
```diff
@@ -739,6 +739,61 @@ fn format_tuple_struct(context: &RewriteContext,
...
+                                                 generics_indent,
+                                                 generics_span));
+
+    result.push_str(&generics_str);
+    result.push_str(" = ");
+
+    let last_line_length = match generics_str.rfind("\n") {
+        Some(index) => " = ".len() + generics_str.len() - index,
+        None => result.len(),
+    };
```
Comment (nrc) [inline src/items.rs]: You can use `utils::last_line_width` for this

File: src/items.rs [hunk 5]
```diff
@@ -739,6 +739,61 @@ fn format_tuple_struct(context: &RewriteContext,
...
+
+    let budget = try_opt!(context.config
+                                 .max_width
+                                 .checked_sub(indent.width() + last_line_length + ";".len()));
+    let type_indent = indent + last_line_length;
+    // Try to fit the type on the same line
+    let ty_str = try_opt!(ty.rewrite(context, budget, type_indent)
+                            .or_else(|| {
+                                // The line was to short try and put the type on the next line
```
Comment (nrc) [inline src/items.rs]: typo: too

**Comment by nrc** [comment]
Awesome, looks great! r+ with the two nits addressed

**Comment by Marwes** [comment]
Messed up the last comment a bit, fixing in a sec.

**Comment by Marwes** [comment]
Done, reworded the last comment a bit.

**Comment by nrc** [comment]
Thanks for the changes!

**Pull request merged**